### PR TITLE
Patch search

### PR DESCRIPTION
**Pull request by jerryzhu509**
Took care of null exceptions in searching (happens when one can't find plugins.)
Improved taking commands: added abbreviations "-n" "-k" "-c" to specify search types.
Added tests: the general idea was to test 2 searches for each type - one that has results and one that is empty.

### CODE REVIEW

File: src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByKeyword.java [hunk 1]
```diff
@@ -63,11 +63,13 @@ public List<PluginModel> getSearchList(String inp, List<PluginModel> plugins) {
         if (SearcherByKeyword.keywordMap == null) {
             SearcherByKeyword.keywordMap = constructSearchMaps(plugins);
         }
-        String [] keywords = inp.toLowerCase().split(" "); // Should be a string
-        Set<PluginModel> res = new HashSet<>(SearcherByKeyword.keywordMap.get(keywords[0]));
+        String [] keywords = inp.split(" "); // Should be a string
+        Set<PluginModel> res = new HashSet<>(SearcherByKeyword.keywordMap.getOrDefault(keywords[0], List.of()));
```
Comment (hykilpikonna) [inline src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByKeyword.java:67]: why was `.toLowerCase()` moved outside this function? the keywords in the map are lowercase though...

**Review by hykilpikonna — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByKeyword.java (hunk 1 above)
Comment (jerryzhu509) [reply src/main/java/org/hydev/mcpm/client/database/searchusecase/SearcherByKeyword.java:67]: should be redundant bc the parser already converts inp to lowercase. Ill add the lower case conversion in searcher again to be safe.

**Review by hykilpikonna — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java [hunk 2]
```diff
@@ -30,20 +28,31 @@ public String description() {
...
     public void configure(Subparser parser) {
-        parser.addArgument("by").choices("name", "keyword", "command")
-                .setDefault("name").dest("type")
-                .help("Specifies the Type of Search");
+        // Default search by name, if -k is specified then search by keyword
+        var type = parser.addMutuallyExclusiveGroup();
+        type.addArgument("-k", "--keyword").action(Arguments.storeTrue()).dest("byKeyword")
+            .help("Search by keyword in descriptions");
+        type.addArgument("-c", "--command").action(Arguments.storeTrue()).dest("byCommand")
+            .help("Search by command");
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java:36]: 👍

File: src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java [hunk 3]
```diff
@@ -30,20 +28,31 @@ public String description() {
...
-        var t = details.getList("text");
-        controller.searchPackages(details.getString("type"),
-                String.join(" ", details.getList("text")),
-                !details.getBoolean("noCache"), log);
+        // Convert type
+        var type = "name";
+        if (details.getBoolean("byKeyword"))
+            type = "keyword";
+        if (details.getBoolean("byCommand"))
+            type = "command";
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/parsers/SearchParser.java:52]: You could probably have multiple arguments with the same `dest` and different `.store*` actions to get a cleaner solution, although I don't totally trust that what I'm suggesting won't catch on fire :P

File: src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java [hunk 4]
```diff
@@ -42,6 +67,36 @@ public static PluginYml meta(String name, String version, String description) {
...
+    /**
+     * Creates a mock PluginYml object.
+     *
+     * @param name The name of the plugin.
+     * @param version The version string for the plugin.
+     * @param description The description for the plugin.
+     * @param commands Commands for the plugin.
+     * @return A PluginYml object.
+     */
+    public static PluginYml meta(String name, String version, String description,
```
Comment (1whatleytay) [inline src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java:51]: You should make the above `meta` function call this one. The above one is less specialized (just pass null to this guy!)

File: src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java [hunk 5]
```diff
@@ -54,6 +109,21 @@ public static PluginVersion version(long id, String name, String string) {
...
+     * Creates a mock PluginVersion object.
+     *
+     * @param id The version id.
+     * @param name The plugin name (for meta).
+     * @param string The version string (for meta).
+     * @param description The plugin description.
+     * @param commands The commands in the plugin.
+     * @return A PluginVersion object.
+     */
+    public static PluginVersion version(long id, String name, String string, String description,
```
Comment (1whatleytay) [inline src/test/java/org/hydev/mcpm/client/database/PluginMockFactory.java:94]: Same deal here.

File: src/test/java/org/hydev/mcpm/client/database/SearchInteractorTest.java [hunk 6]
```diff
@@ -39,39 +48,85 @@ private String formatStr(SearchPackagesResult result, String delim) {
...
+        var text = formatStr(result, ", ");
+        assert text.equals("WorldGuard, Holographic Displays");
+    }
+
+    @Test
+    void testSearchByCommandSuccessNoMatch() {
+        var result = database.search(
+                new SearchPackagesInput(SearchPackagesType.BY_COMMAND, "pp", true));
+
+        assert result.state() == SearchPackagesResult.State.SUCCESS;
```
Comment (1whatleytay) [inline src/test/java/org/hydev/mcpm/client/database/SearchInteractorTest.java]: `assertEquals(...)` please. There's a few of these regular asserts around this file.

**Review by 1whatleytay — APPROVED** [review]
Nice tests!

**Pull request merged**